### ISSUE TITLE
Range is not balancing across the cluster

### ISSUE THREAD

**Issue by gslin** [cockroachdb/cockroach#33599]
**What is your situation?**

Range is not balancing across the cluster.

We were running CockroachDB 2.1.3 benchmarking on Kubernetes with 10 EC2 c5.large nodes (running CockroachDB) and another EC2 c5.2xlarge node (running sysbench).  All systems (hosts & containers) are based on Ubuntu 18.04 (x86_64) and latest stable software.

Running CockroachDB with:

```bash
cockroach start --background --insecure --listen-addr=0.0.0.0 --advertise-addr=$(ifconfig eth0 | grep inet | awk '{print $2}')
cockroach start --background --insecure --listen-addr=0.0.0.0 --join=192.168.x.x --advertise-addr=$(ifconfig eth0 | grep inet | awk '{print $2}')
```

Change `num_replica` to 3 and create database:

```sql
ALTER RANGE default CONFIGURE ZONE USING num_replicas = 3;
CREATE DATABASE sbtest;
```

Benchmark with sysbench:
```bash
sysbench --db-driver=pgsql --pgsql-host=cockroachdb-port.default.svc.cluster.local --pgsql-port=26257 --pgsql-user=root --pgsql-password= --time=0 --events=10000000 --report-interval=1 --threads=128 /usr/share/sysbench/oltp_insert.lua prepare
sysbench --db-driver=pgsql --pgsql-host=cockroachdb-port.default.svc.cluster.local --pgsql-port=26257 --pgsql-user=root --pgsql-password= --time=0 --events=10000000 --report-interval=1 --threads=128 /usr/share/sysbench/oltp_insert.lua run
```

I ran it about 6~7 times and got https://gist.github.com/gslin/93acb2da748a88cd09e25f06e3a01c38 this.

As you can see, `lease_holder` is not balanced:

```
$ awk '{print $NF}' cockroachdb-range-2.txt | sort | uniq -c | sort
     10 1
     15 3
     19 5
     31 9
     43 4
     46 2
     49 10
     49 8
     53 6
     59 7
```

**Requested resolution**

Range should be balancing across the cluster.

**Comment by knz** [comment]
@tim-o could you qualify?

**Comment by tim-o** [comment]
Hi @gslin -it looks like you're using ints as your primary key. We'd expect hotspots unless you're using uuids. How do the results look for test on a schema with UUIDs as the primary key?

**Comment by gslin** [comment]
Okay, I will try it with different PK schema.

So is this a known issue for interger PK? I think it's quite common about using integer on PK.

**Comment by gslin** [comment]
Thanks, I just updated CockroachDB version information (it's 2.1.3).

And about the workload, it's sysbench's script https://github.com/akopytov/sysbench/blob/master/src/lua/oltp_insert.lua and you can see it's just a simple INSERT loop, no SELECT and others.

**Comment by knz** [comment]
"A simple INSERT loop" - wait, not so fast. This logic is not that simple because what it precisely does is highly dependent on the parameters you enter in the sysbench configuration.

1) Depending on the `auto_inc` parameter:

   - if enabled, the table will use SERIAL, which is equivalent to `unique_rowid()` and thus subject to the contention risk described here: https://www.cockroachlabs.com/docs/stable/sql-faqs.html#what-are-the-differences-between-uuid-sequences-and-unique_rowid
   - if disabled, it will produce random 32-bit values, which is *much* better than `unique_rowid()` but is still likely to cause contention in cockroachdb. We recommend 64 bit random values instead (or better yet UUID, like initially hinted by Tim).

2) the benchmark either writes to just 1 table, or to many tables (configurable parameter). What parameter did you choose? This influences our understanding of your problem.

3) This workload uses a single value on each INSERT, which will make the performance results artificially lower - see https://www.cockroachlabs.com/docs/stable/insert.html#insert-multiple-rows-into-an-existing-table

In short, I suspect you have the `auto_inc` parameter enabled and are seeing the effect of many of the INSERT queries hitting the same range over and over, which may impact leaseholder rebalancing negatively. If that is the case, I'd recommend you disable the `auto_inc` parameter in sysbench.

**Comment by dourvaris** [comment]
I've noticed the same balancing issue as well, and have done some investigation here: https://gist.github.com/dourvaris/0176277c1d0059ff433cdc185b14266d - the interesting part is going from 3 => 6 => 3 replicas leaves some ranges at 4+ replicas on the first scatter - and in all cases, the balance of the holders is skewed - the example uses 24 ranges, but in practice with a table of 1000s of ranges the behavior is identical

Investigation done on 6 node cockroachdb/cockroach (tag:v2.1.3) cluster on a single `create table test (id uuid primary key)`

**Comment by knz** [comment]
@petermattis who is best to triage this?

**Comment by gslin** [comment]
@petermattis Only one database (`sbtest`) and one table (`sbtest1`).
